sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.scala

```diff
@@ -0,0 +1,408 @@
+
+  /**
+   * ColumnarBatch for vectorized execution by whole-stage codegen.
+   */
+  private var columnarBatch: ColumnarBatch = _
+
+  /**
+   * Writable columnVectors of ColumnarBatch.
+   */
+  private var columnVectors: Seq[WritableColumnVector] = _
+
+  /**
+   * The number of rows read and considered to be returned.
+   */
+  private var rowsReturned: Long = 0L
+
+  /**
+   * Total number of rows.
+   */
+  private var totalRowCount: Long = 0L
+
+  override def getCurrentKey: Void = null
+
+  override def getCurrentValue: ColumnarBatch = columnarBatch
+
+  override def getProgress: Float = rowsReturned.toFloat / totalRowCount
+
+  override def nextKeyValue(): Boolean = nextBatch()
+
+  override def close(): Unit = {
+    if (columnarBatch != null) {
+      columnarBatch.close()
+      columnarBatch = null
+    }
+    if (recordReader != null) {
+      recordReader.close()
+      recordReader = null
+    }
+  }
+
+  /**
+   * Initialize ORC file reader and batch record reader.
+   * Please note that `setRequiredSchema` is needed to be called after this.
+   */
+  override def initialize(inputSplit: InputSplit, taskAttemptContext: TaskAttemptContext): Unit = {
+    val fileSplit = inputSplit.asInstanceOf[FileSplit]
+    val conf = taskAttemptContext.getConfiguration
+    reader = OrcFile.createReader(
+      fileSplit.getPath,
+      OrcFile.readerOptions(conf)
+        .maxLength(OrcConf.MAX_FILE_LENGTH.getLong(conf))
+        .filesystem(fileSplit.getPath.getFileSystem(conf)))
+
+    val options = OrcInputFormat.buildOptions(conf, reader, fileSplit.getStart, fileSplit.getLength)
+    recordReader = reader.rows(options)
+  }
+
+  /**
+   * Set required schema and partition information.
+   * With this information, this creates ColumnarBatch with the full schema.
+   */
+  def setRequiredSchema(
+      orcSchema: TypeDescription,
+      requestedColIds: Array[Int],
+      resultSchema: StructType,
+      requiredSchema: StructType,
+      partitionValues: InternalRow): Unit = {
+    batch = orcSchema.createRowBatch(DEFAULT_SIZE)
+    assert(!batch.selectedInUse)
+    totalRowCount = reader.getNumberOfRows
+    logDebug(s"totalRowCount = $totalRowCount")
+
+    this.requiredSchema = requiredSchema
+    this.requestedColIds = requestedColIds
+
+    val capacity = ColumnarBatch.DEFAULT_BATCH_SIZE
+    if (DEFAULT_MEMORY_MODE == MemoryMode.OFF_HEAP) {
+      columnVectors = OffHeapColumnVector.allocateColumns(capacity, resultSchema)
+    } else {
+      columnVectors = OnHeapColumnVector.allocateColumns(capacity, resultSchema)
+    }
+    columnarBatch = new ColumnarBatch(resultSchema, columnVectors.toArray, capacity)
+
+    if (partitionValues.numFields > 0) {
+      val partitionIdx = requiredSchema.fields.length
+      for (i <- 0 until partitionValues.numFields) {
+        ColumnVectorUtils.populate(columnVectors(i + partitionIdx), partitionValues, i)
+        columnVectors(i + partitionIdx).setIsConstant()
+      }
+    }
```

Comment:
shall we handle missing columns like the parquet reader does?

Reply:
Sure!